torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1071,51 +1070,104 @@ def _get_param_to_param_id_from_optim_input(
+    r0_param_id_to_optim_state_key: Dict[
+        int, _OptimStateKey
+    ] = {}  # rank 0
+
+    for param_id, param in enumerate(param_id_to_param):
+        # Do not include parameters without state to avoid empty mappings
+        # just like in normal `torch.optim.Optimizer.state_dict()`
+        if param_id not in optim_state_dict["state"]:
+            continue
+        optim_state_key = _OptimStateKey(
+            unflat_param_names=tuple(param_to_fqns[param]),
+            is_flat_param=isinstance(param, FlatParameter),
+        )
+        if rank == 0:
+            r0_param_id_to_optim_state_key[param_id] = optim_state_key
+        optim_state_key_to_param_id[optim_state_key] = param_id
+    key_obj_list: List[Optional[Dict[int, _OptimStateKey]]] = (
+        [r0_param_id_to_optim_state_key] if rank == 0 else [None]
     )
-    return [
-        unflat_to_flat_param_ids[unflat_param_id]
-        for unflat_param_id in range(num_unflat_param_ids)
-    ]
+    dist.broadcast_object_list(key_obj_list, src=0, group=group)
+    assert key_obj_list[0] is not None
+    r0_param_id_to_optim_state_key = key_obj_list[0]
 
+    # Ensure that all ranks have at least the optimizer states needed by
+    # rank 0's optimizer
+    missing_keys: List[_OptimStateKey] = []
+    for r0_optim_state_key in r0_param_id_to_optim_state_key.values():
+        if r0_optim_state_key not in optim_state_key_to_param_id:
+            # A parameter from rank 0's optimizer does not exist for this
+            # rank's optimizer
+            missing_keys.append(r0_optim_state_key)
+            continue
+        param_id = optim_state_key_to_param_id[r0_optim_state_key]
+        assert param_id >= 0 and param_id < len(
+            param_id_to_param
+        ), "Check the `param_id_to_param` construction"
+    device = torch.device("cuda", torch.cuda.current_device())
+    num_missing = torch.tensor([len(missing_keys)], dtype=torch.int32, device=device)
+    dist.all_reduce(num_missing, group=group)
+    if num_missing.item() > 0:
+        obj_list = [None for _ in range(dist.get_world_size(group))]
+        dist.all_gather_object(obj_list, missing_keys, group=group)
+        error_msg = (
+            "FSDP currently requires each rank to have at least the "
+            "optimizer states needed by rank 0's optimizer but some ranks "
+            "are missing some of those states"
+        )
+        for rank, keys in enumerate(obj_list):
+            keys = cast(List[_OptimStateKey], keys)
+            if len(keys) > 0:
+                error_msg += (
+                    f"\nRank {rank} is missing states for the parameters: "
+                    f"{[key.unflat_param_names for key in keys]}"
+                )
+        raise RuntimeError(error_msg)
 
-def _is_zero_dim_tensor(x: Any) -> bool:
-    return torch.is_tensor(x) and x.dim() == 0
+    return r0_param_id_to_optim_state_key, optim_state_key_to_param_id
+
+
+def _process_param_groups(
```

Comment:
nit: I am wondering, should we call this `_unflatten_param_groups` to be more specific?